File: apps/site/scripts/generate-favicon.py
"""
Regenerates favicon/touch-icon/navbar assets from public/arkitect-mark.png
(the master transparent mark). Run after replacing that master file:

    python scripts/generate-favicon.py

Small badge sizes (16/32) get a brightness/saturation boost so the mark
reads clearly against the dark badge background at tiny scale; the
apple-touch-icon and navbar asset use the mark as-is.
"""
import io
import os

import numpy as np
from PIL import Image, ImageDraw, ImageEnhance

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PUBLIC = os.path.join(ROOT, "public")
MASTER_PATH = os.path.join(PUBLIC, "arkitect-mark.png")
DARK_BG = (17, 21, 33, 255)  # #111521 — site's dark brand surface


def brighten_for_small(im, gamma=0.5, sat=1.35):
    arr = np.asarray(im).astype(np.float32)
    rgb = np.power(np.clip(arr[..., :3], 0, 255) / 255.0, gamma) * 255.0
    out = np.dstack([rgb, arr[..., 3]]).astype(np.uint8)
    result = Image.fromarray(out, mode="RGBA")
    r, g, b, a = result.split()
    boosted = ImageEnhance.Color(Image.merge("RGB", (r, g, b))).enhance(sat)
    r2, g2, b2 = boosted.split()
    return Image.merge("RGBA", (r2, g2, b2, a))


def rounded_mask(size, radius):
    mask = Image.new("L", (size, size), 0)
    ImageDraw.Draw(mask).rounded_rectangle([0, 0, size - 1, size - 1], radius=radius, fill=255)
    return mask


def badge(size, fill_frac, radius_frac, source):
    canvas = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    bg_layer = Image.new("RGBA", (size, size), DARK_BG)
    canvas = Image.composite(bg_layer, canvas, rounded_mask(size, int(size * radius_frac)))
    inner = int(size * fill_frac)
    mark = source.resize((inner, inner), Image.LANCZOS)
    off = (size - inner) // 2
    canvas.alpha_composite(mark, (off, off))
    return canvas


def png_to_ico(png_buffers, sizes, out_path):
    count = len(png_buffers)
    header = bytearray(6)
    header[2] = 1
    header[4] = count & 0xFF
    entries = bytearray()
    offset = 6 + count * 16
    for buf, size in zip(png_buffers, sizes):
        entry = bytearray(16)
        entry[0] = 0 if size >= 256 else size
        entry[1] = 0 if size >= 256 else size
        entry[4] = 1
        entry[6] = 32
        entry[8:12] = len(buf).to_bytes(4, "little")
        entry[12:16] = offset.to_bytes(4, "little")
        entries += entry
        offset += len(buf)
    with open(out_path, "wb") as f:
        f.write(header)
        f.write(entries)
        for buf in png_buffers:
            f.write(buf)


def main():
    master = Image.open(MASTER_PATH).convert("RGBA")
    bright = brighten_for_small(master)

    fav32 = badge(32, 0.86, 0.20, bright)
    fav16 = badge(16, 0.92, 0.22, bright)
    fav32.save(os.path.join(PUBLIC, "favicon-32.png"), optimize=True)
    fav16.save(os.path.join(PUBLIC, "favicon-16.png"), optimize=True)

    buf16, buf32 = io.BytesIO(), io.BytesIO()
    fav16.save(buf16, format="PNG")
    fav32.save(buf32, format="PNG")
    png_to_ico([buf16.getvalue(), buf32.getvalue()], [16, 32], os.path.join(PUBLIC, "favicon.ico"))

    apple = Image.new("RGBA", (180, 180), DARK_BG)
    inner = int(180 * 0.72)
    apple.alpha_composite(master.resize((inner, inner), Image.LANCZOS), ((180 - inner) // 2,) * 2)
    apple.convert("RGB").save(os.path.join(PUBLIC, "apple-touch-icon.png"), optimize=True)

    nav = master.resize((128, 128), Image.LANCZOS)
    nav.save(os.path.join(PUBLIC, "arkitect-mark-nav.png"), optimize=True)

    print("Wrote favicon-16.png, favicon-32.png, favicon.ico, apple-touch-icon.png, arkitect-mark-nav.png")


if __name__ == "__main__":
    main()

File: apps/site/scripts/generate-favicon.mjs
import { readFileSync, writeFileSync, unlinkSync } from "node:fs";
import { spawnSync } from "node:child_process";
import path from "node:path";
import { fileURLToPath } from "node:url";

const root = path.dirname(fileURLToPath(import.meta.url));
const publicDir = path.join(root, "..", "public");
const svgPath = path.join(publicDir, "favicon.svg");
const png16 = path.join(publicDir, "favicon-16.png");
const png32 = path.join(publicDir, "favicon-32.png");
const icoPath = path.join(publicDir, "favicon.ico");

function renderPng(size, output) {
  const result = spawnSync(
    "npx",
    ["--yes", "@resvg/resvg-js-cli", "--fit-width", String(size), svgPath, output],
    { stdio: "inherit", shell: true }
  );
  if (result.status !== 0) {
    throw new Error(`Failed to render ${size}px PNG`);
  }
}

function pngToIco(pngBuffers, sizes) {
  const count = pngBuffers.length;
  const header = Buffer.alloc(6);
  header.writeUInt16LE(0, 0);
  header.writeUInt16LE(1, 2);
  header.writeUInt16LE(count, 4);

  const entries = [];
  let offset = 6 + count * 16;

  for (let i = 0; i < count; i += 1) {
    const png = pngBuffers[i];
    const size = sizes[i];
    const entry = Buffer.alloc(16);
    entry.writeUInt8(size === 256 ? 0 : size, 0);
    entry.writeUInt8(size === 256 ? 0 : size, 1);
    entry.writeUInt8(0, 2);
    entry.writeUInt8(0, 3);
    entry.writeUInt16LE(1, 4);
    entry.writeUInt16LE(32, 6);
    entry.writeUInt32LE(png.length, 8);
    entry.writeUInt32LE(offset, 12);
    entries.push(entry);
    offset += png.length;
  }

  return Buffer.concat([header, ...entries, ...pngBuffers]);
}

renderPng(16, png16);
renderPng(32, png32);

const png16Buf = readFileSync(png16);
const png32Buf = readFileSync(png32);
writeFileSync(icoPath, pngToIco([png16Buf, png32Buf], [16, 32]));

unlinkSync(png16);
unlinkSync(png32);

console.log(`Wrote ${icoPath} (${readFileSync(icoPath).length} bytes)`);
